BB-727: Fix ZookeeperConfigManager event listener leaks

BourgoisMickael opened the pull request:
- improve some logs
- warn early for event listener leaks
- avoid send a remove to zk when a watcher notify the removed bucket config
- separate list and get configs

Below a sequence diagram showing in red a few problems fixed here

<img width="1696" height="1601" alt="bucket-notif-zookeeper drawio" src="https://github.com/user-attachments/assets/924d6b5d-6ceb-4428-9f37-1fafba5ce009" />

BourgoisMickael committed:
BB-727: Fix comment typo

BourgoisMickael committed:
BB-727: Better print buffer in logs

Fix `key: { type: 'Buffer', data: [ 104, 101, 108, 108, 111 ] }`
To `key: 'hello'`

BourgoisMickael committed:
BB-727: removeConfig only locally on NODE_DELETED

The watcher NODE_DELETED already calls removeConfig with false
Because it's already removed from zk if the event is received
There is no need to send a request to zk

BourgoisMickael committed:
BB-727: Log bucket for invalid config error

This kind of error can happen if data is undefined
when the read is triggered too late when the znode
is already removed

BourgoisMickael committed:
BB-727: Check and warn event emitter leak early

Otherwise we need a few buckets and bucket operations
to trigger the event emitter listener limit warning at 10 listeners

BourgoisMickael committed:
BB-727: Log as debug the list of buckets

BourgoisMickael committed:
BB-727: Separate list and get configs

The getConfig should not trigger a listing with watcher
as it duplicates an already existing watcher.
This can have an exponential effect of adding even more watchers
when events are triggered.

A clear separation prevent this effect:
- only NODE_CHILDREN_CHANGED should trigger a new listing to reapply a watcher
- getConfig should be contained to a single bucket (discovered by listing, or from watcher NODE_DATA_CHANGED)

bert-e commented:
# Hello bourgoismickael,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

BourgoisMickael committed:
BB-727: Watch created config in queuePopulator

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-727 contains:


* `9.0.21`

* `9.1.5`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.22`

* `9.1.5`


Please check the `Fix Version/s` of BB-727, or the target
branch of this pull request.

BourgoisMickael committed:
BB-727: Improve mkdirp to create znode with data

Atomic mkdirp to prevent trigger a watcher for znode creation
before data is set on the znode

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

BourgoisMickael commented:
/create_integration_branches

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-727-zk-config` into
`development/9.0`
* [w/9.1/improvement/BB-727-zk-config](https://github.com/scality/backbeat/compare/development/9.1...w/9.1/improvement/BB-727-zk-config) will be merged into `development/9.1` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

BourgoisMickael committed:
BB-727: Update tests to use ZookeeperManager

BourgoisMickael committed:
BB-727: Unit test mkdirpWithChildDataOnly

bert-e commented:
# History mismatch



Merge commit #7edd3977408e2654642035c90a030202a5b764bc on the integration branch
`w/9.1/improvement/BB-727-zk-config` is merging a branch which is neither the current
branch `improvement/BB-727-zk-config` nor the development branch
`development/9.1`.

It is likely due to a rebase of the branch `improvement/BB-727-zk-config` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**





*The following options are set:* **create_integration_branches**

BourgoisMickael commented:
@bert-e reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **create_integration_branches**

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/BB-727-zk-config` into
`development/9.0`
* [w/9.1/improvement/BB-727-zk-config](https://github.com/scality/backbeat/compare/development/9.1...w/9.1/improvement/BB-727-zk-config) will be merged into `development/9.1` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**
